Port GAPIC metadata generation for Python

geigerj opened the pull request:
Move the functionality of generating Python GAPIC package
metadata from Packman to GAPIC. To do so, introduce a new package
metadata configuration format, which lives at the same level as the
GAPIC config in terms of input/test input. This new config file
fixes #848.

Sorry, this PR is big, but a lot of the LOC changed are just test data.

The following follow-up PRs are planned:
- A PR to Artman, introducing a task to generate the
  new package metadata config from the artman config, and removing the
  Python Packman task.
- A PR to Toolkit, replacing all inputs to GrpcMetadataGen with the
  new package metadata config, which addresses #780
- (Potentially) A PR to Toolkit investigating the consolidation of
  gRPC and GAPIC Python package metadata snippet templates. Note that
  Packman currently uses separate templates for gRPC and GAPIC.

cc: @bjwatson, @shinfan

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/config/PackageMetadataConfig.java:
```diff
@@ -0,0 +1,203 @@
+public abstract class PackageMetadataConfig {
+
+  @AutoValue
+  public abstract static class VersionBound {
```
It seems like this concept could be usable outside of PackageMetadataConfig - maybe lift up to its own file?

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/config/PackageMetadataConfig.java:
```diff
@@ -0,0 +1,203 @@
+    public abstract String upper();
+  }
+
+  protected abstract Map<String, VersionBound> gaxVersion();
```
The name here (`gaxVersion`) implies a single version; maybe `gaxVersionBound` would be clearer. (Applies to the attributes below too)

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/config/PackageMetadataConfig.java:
```diff
@@ -0,0 +1,203 @@
+  }
+
+  @SuppressWarnings("unchecked")
+  public static PackageMetadataConfig createFromFile(Path packageConfigPath) throws IOException {
```
This seems like transformer logic

shinfan replied:
We usually put such logic into a separate class and this package config class only a container class.

geigerj replied:
Not sure I understand -- are you talking about the `*PackageMetadataTransformer` classes? Why would they be concerned with file I/O?

shinfan replied:
I don't think transformers should handle I/O. It is probably okay to keep it here then.

garrettjonesgoogle replied:
I think the actual place where file I/O should be handled is at the top level; transformers accept an in-memory data model as input.

geigerj replied:
Moved I/O out of this method. Do you still think this should be moved to the transformer? Currently, this is the standin for the (hidden) builder, since there's no one that currently needs to build this class explicitly.

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/config/PackageMetadataConfig.java:
```diff
@@ -0,0 +1,203 @@
+  protected abstract Map<String, String> packageName();
+
+  /** The version of GAX that this package depends on. Configured per language. */
+  public VersionBound gaxVersion(String language) {
```
I'd prefer not to pass around language identifiers as plain strings - lots of room for odd variations & mistakes, and it is an "almost" closed set.

geigerj replied:
Created an enum for this

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/py/PythonPackageMetadataTransformer.java:
```diff
@@ -0,0 +1,182 @@
+              public ViewModel apply(@Nullable String templateFileName) {
+                return generateMetadataView(model, apiConfig, templateFileName);
+              }
+            });
```
This would be much more concise with a simple for loop.

geigerj replied:
Done.

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/py/PythonPackageMetadataTransformer.java:
```diff
@@ -0,0 +1,182 @@
+        first = false;
+      }
+      packages.add(current.toString());
+    }
```
At first read, it looks like you're splitting by "." and then joining by "."... what is this supposed to do?

geigerj replied:
A Python package name `foo.bar.baz` implicitly defines three packages: `foo`, `foo.bar`, and `foo.bar.baz`. Is there a more concise way to express this in Java?

garrettjonesgoogle replied:
My fault for not reading the comment. I didn't realize the accumulative aspect of the algorithm. 

Here is an alternate approach (not checked for correctness - beware off-by-one errors):

```
List<String> parts = Splitter.on(".").split(packageName);
for (int i = 1; i <= parts.size(); i++) {
  packages.add(Joiner.on(".").join(parts.subList(0, i)));
}
```

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/py/PythonPackageMetadataTransformer.java:
```diff
@@ -0,0 +1,182 @@
+              public boolean apply(@Nullable String input) {
+                return isNamespacePackage(input, apiVersion);
+              }
+            }));
```
This would be simpler as a for loop.

geigerj replied:
Fewer Java lines, maybe, but simpler? This situation is what a filter is *for*...

(but anyhow, done)

garrettjonesgoogle replied:
Filters should look much nicer in Java 8 when it rolls around... but in Java 7, they are pretty ugly.

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/py/PythonPackageMetadataTransformer.java:
```diff
@@ -0,0 +1,182 @@
+            } else {
+              template = "py/__init__.py.snip";
+            }
+            return new ViewModel() {
```
Just make a simple ViewModel class with AutoBuilder that can be used here. It feels kind of odd to construct an anonymously-typed object for every instance.

geigerj replied:
Done

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/viewmodel/PackageMetadataView.java:
```diff
@@ -33,15 +36,36 @@ public String resourceRoot() {
   public abstract String identifier();
 
-  public abstract String version();
+  public abstract VersionBound packageVersion();
```
Similar to my comment within PackageMetadataConfig: +`Bound`

geigerj replied:
Done

garrettjonesgoogle commented on src/main/resources/com/google/api/codegen/nodejs/package.snip:
```diff
@@ -1,23 +1,23 @@
       "mocha": "3.2.0"
     },
-    "license": "Apache-2.0",
+    "license": "{@metadata.license}",
```
The name `license` implies the content of the license; maybe this should be `licenseName`.

geigerj replied:
Done

shinfan commented on src/main/java/com/google/api/codegen/config/PackageMetadataConfig.java:
```diff
@@ -0,0 +1,203 @@
+
+  protected abstract Map<String, VersionBound> gaxVersion();
+
+  protected abstract Map<String, VersionBound> protoVersion();
```
What is the key of this map?

geigerj replied:
Changed to `TargetLanguage` enum

shinfan commented on src/main/java/com/google/api/codegen/gapic/GapicProviderFactory.java:
```diff
@@ -15,12 +15,16 @@
+      Model model,
+      ApiConfig apiConfig,
+      GapicGeneratorConfig generatorConfig,
+      PackageMetadataConfig packageConfig);
```
Does it make sense to put PackageMetadataConfig under GapicGeneratorConfig? I am a bit cautious of adding more parameters to this interface and conceptually package metadata is a part of GAPIC gen now.

shinfan replied:
Fair enough. Could you update the comment o the `ApiConfig` class to point out that it represents the contents of {api}_gapic.yaml. Thanks.

geigerj replied:
Done.

shinfan commented on src/main/java/com/google/api/codegen/transformer/py/PythonPackageMetadataTransformer.java:
```diff
@@ -0,0 +1,182 @@
+import java.util.List;
+import javax.annotation.Nullable;
+
+public class PythonPackageMetadataTransformer implements ModelToViewTransformer {
```
nit: add comments

shinfan commented on src/main/java/com/google/api/codegen/transformer/py/PythonPackageMetadataTransformer.java:
```diff
@@ -0,0 +1,182 @@
+        .build();
+  }
+
+  private List<String> computePackages(String packageName) {
```
Add a comment of the computing logic here and `computeInitFiles`

shinfan commented on src/main/resources/com/google/api/codegen/common.snip:
```diff
@@ -33,3 +33,9 @@
+    @join element : list on ", "
+        '{@element}'
+    @end
+@end
```
nit: new line

geigerj replied:
Done

shinfan commented on src/test/java/com/google/api/codegen/testdata/library_pkg.yaml:
```diff
@@ -0,0 +1,48 @@
+
+# Language-dependent configuration
+package_name:
+  default: google-cloud-library-v1
```
Add a language specific package name here for better coverage

geigerj replied:
Respectfully disagree: there's coverage for language-specific configuration elsewhere in this file. This tests the logic for handling the "default" case.

shinfan replied:
Oh..... you are right.  Default case here should be better.

geigerj commented on src/main/java/com/google/api/codegen/transformer/nodejs/NodeJSPackageMetadataTransformer.java:
```diff
@@ -45,19 +45,29 @@
+        .gaxVersion(packageConfig.gaxVersion("nodejs"))
+        .protoVersion(packageConfig.protoVersion("nodejs"))
+        .commonProtosVersion(packageConfig.commonProtosVersion("nodejs"))
+        .packageName(packageConfig.packageName("nodejs"))
```
What would you suggest here? Is it sufficient that this to be changed to an enum, or are you specifically concerned with the repetition? (I think it's probably overkill to create a per-language "view" of the config object.)

geigerj commented:
@garrettjonesgoogle @shinfan PTAL -- I addressed most of the comments. Would appreciate more guidance the MVVM stuff regarding resource files vs. templates and whether transformers should be doing I/O.

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/config/PackageMetadataConfig.java:
```diff
@@ -0,0 +1,205 @@
+  public static PackageMetadataConfig createFromFile(Path packageConfigPath) throws IOException {
+    Yaml yaml = new Yaml();
+    String contents = new String(Files.readAllBytes(packageConfigPath), StandardCharsets.UTF_8);
+    Map<String, Object> configMap = (Map<String, Object>) yaml.load(contents);
```
All of our other yaml uses proto as a schema - is it worth converting this to use the same thing?

geigerj replied:
I'm not sure see the benefit, other than for consistency. Creating a proto schema increases the number of places you need to touch if you want to change the schema, and it forces everyone to rebuild the generated Java classes. Given that the schema here is likely to be quite simple, do you think this is worthwhile?

garrettjonesgoogle replied:
I see the value as having nice type safety over a configuration model. If you have no schema, your validation is at run time / test time. It might not be worthwhile here if it's simple enough, though.

garrettjonesgoogle replied:
Here is the difference:

direct yaml loading:

1. Validation of config file: validated at runtime when build() runs
2. Validation of code accessing data from the config file: validated at runtime when build() runs

Note - you can't distinguish between 1 & 2. 

proto-based loading:

1. Validation of config file: validated at runtime in MultiYamlReader
2. Validation of code accessing data from the config file: validated at compile time against Java classes generated by the schema

Like I said, this is simple enough that I won't require a proto-based schema for it. I just wanted to make sure we agree on the tradeoffs.

geigerj replied:
Yup, agreed. I see what you're saying -- the tradeoff is in the type safety of the (relatively simple) code in `createFromFile` here as opposed to the (relatively complex) code in, say, `createApiConfig`.

garrettjonesgoogle commented on src/main/java/com/google/api/codegen/transformer/py/PythonPackageMetadataTransformer.java:
```diff
@@ -0,0 +1,161 @@
+        .namespacePackages(
+            computeNamespacePackages(apiConfig.getPackageName(), packageConfig.apiVersion()))
+        .hasMultipleServices(false)
+        .build();
```
How much of this is the same between Python & Node? (It's hard to eyeball it)

garrettjonesgoogle replied:
If the differences are that small, you could make a common transformer and set slightly more values than any particular language needs.

geigerj replied:
That's true, but I think it's more readable to keep it this way than to have a class that incorporates a `NodeJSPackageMetadataNamer`, some bits of logic about Python namespace packages, and presumably soon other small bits of other language-specific logic. WDYT?

garrettjonesgoogle replied:
This code is kind of a long stretch to copy for every language...

I don't think you need a metadata namer. You could just pass in TargetLanguage as an argument.

Here are some ideas for strategies for setting language-specific bits:

1. Have the transformer take an empty Builder for PackageMetadataView instead of creating its own fully-constructed instance; the caller can then fill in its own bits here & there and call build() itself.
2. Pass in the extra language-specific things as arguments. (This unfortunately is likely to grow)
3. Create language-specific specialty views: PythonExtraMetadata (holds namespacePackages), NodeExtraMetadata (holds hasMultipleServices, identifier)

geigerj replied:
Done.

garrettjonesgoogle commented on src/main/resources/com/google/api/codegen/LICENSE.snip:
```diff
@@ -0,0 +1,204 @@
+@snippet generate(context)
+    {@""}                            Apache License
```
Hmm if this is accessible from more than one language, then maybe it's fine.

garrettjonesgoogle commented:
LGTM

geigerj commented:
@shinfan Do you want to take another look before I merge?

shinfan commented:
LGTM
